File: backend/core/lexer.py
# backend/core/lexer.py

from enum import Enum


class TokenType(Enum):
    NUMBER = 1
    VARIABLE = 2
    PLUS = 3
    MINUS = 4
    MUL = 5
    DIV = 6
    EQUAL = 7
    LPAREN = 8
    RPAREN = 9


class Token:
    __slots__ = ("type", "value")

    def __init__(self, type_, value=None):
        self.type = type_
        self.value = value

    def __repr__(self):
        return f"{self.type.name}:{self.value}" if self.value is not None else f"{self.type.name}"


# Fast O(1) operator lookup
_OPERATOR_MAP = {
    "+": TokenType.PLUS,
    "-": TokenType.MINUS,
    "*": TokenType.MUL,
    "/": TokenType.DIV,
    "=": TokenType.EQUAL,
    "(": TokenType.LPAREN,
    ")": TokenType.RPAREN,
}


def tokenize(expression: str):
    tokens = []
    append = tokens.append  # local binding for speed

    i = 0
    n = len(expression)

    while i < n:
        char = expression[i]

        # Skip whitespace (fast path)
        if char <= " ":
            i += 1
            continue

        # NUMBER (multi-digit, optimized slicing)
        if "0" <= char <= "9":
            start = i
            i += 1
            while i < n and "0" <= expression[i] <= "9":
                i += 1
            append(Token(TokenType.NUMBER, int(expression[start:i])))
            continue

        # VARIABLE (multi-letter + digits allowed after first char)
        if ("a" <= char <= "z") or ("A" <= char <= "Z"):
            start = i
            i += 1

            while i < n:
                c = expression[i]
                if ("a" <= c <= "z") or ("A" <= c <= "Z") or ("0" <= c <= "9"):
                    i += 1
                else:
                    break

            append(Token(TokenType.VARIABLE, expression[start:i]))
            continue

        # OPERATOR / SYMBOL
        token_type = _OPERATOR_MAP.get(char)
        if token_type:
            append(Token(token_type))
            i += 1
            continue

        # ERROR handling (clear + useful)
        raise ValueError(f"Invalid character '{char}' at position {i}")

    return tokens

File: backend/core/lexer.cpp
#include "lexer.h"
#include <cctype>
#include <stdexcept>

namespace quant {

Lexer::Lexer(std::string input)
    : input_(std::move(input)), position_(0) {}

char Lexer::peek() const {
    return position_ < input_.size() ? input_[position_] : '\0';
}

char Lexer::consume() {
    return position_ < input_.size() ? input_[position_++] : '\0';
}

void Lexer::skipWhitespace() {
    while (std::isspace(static_cast<unsigned char>(peek()))) {
        consume();
    }
}

Token Lexer::nextToken() {
    skipWhitespace();
    char current = peek();
    if (current == '\0') {
        return Token(TokenType::End, "", 0);
    }

    if (std::isdigit(static_cast<unsigned char>(current))) {
        size_t start = position_;
        long long value = 0;
        while (std::isdigit(static_cast<unsigned char>(peek()))) {
            value = value * 10 + (consume() - '0');
        }
        return Token(TokenType::Number, input_.substr(start, position_ - start), value);
    }

    if (std::isalpha(static_cast<unsigned char>(current))) {
        size_t start = position_;
        while (std::isalnum(static_cast<unsigned char>(peek()))) {
            consume();
        }
        return Token(TokenType::Variable, input_.substr(start, position_ - start), 0);
    }

    consume();
    switch (current) {
        case '+': return Token(TokenType::Plus, "+", 0);
        case '-': return Token(TokenType::Minus, "-", 0);
        case '=': return Token(TokenType::Equals, "=", 0);
        default:
            throw std::invalid_argument("Invalid character in expression: " + std::string(1, current));
    }
}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;
    while (true) {
        Token token = nextToken();
        if (token.type == TokenType::End) {
            break;
        }
        tokens.push_back(std::move(token));
    }
    return tokens;
}

} // namespace quant
